kong/router/compat.lua

```diff
@@ -295,8 +302,58 @@ local function get_exp_and_priority(route)
 end
 
 
-function _M.new(routes, cache, cache_neg, old_router)
-  return atc.new(routes, cache, cache_neg, old_router, get_exp_and_priority)
+-- group array-like table t by the function f, returning a table mapping from
+-- the result of invoking f on one of the elements to the actual elements.
+local function group_by(t, f)
+  local result = {}
+  for _, value in ipairs(t) do
+    local key = f(value)
+    if result[key] then
+      result[key][#result[key] + 1] = value
+    else
+      result[key] = {value}
+    end
+  end
+  return result
+end
+
+-- split routes into multiple routes, one for each prefix length and one for all
+-- regular expressions
+local function split_route_by_path_into(rs, split_rs)
+  if is_empty_field(rs.route.paths) or #rs.route.paths == 1 then
+    split_rs[#split_rs + 1] = rs
+    return
+  end
+
+  assert(rs.route.paths)
+  local grouped_paths = group_by(
+    rs.route.paths,
+    function(path)
+      return is_regex_magic(path) or #path
+    end
+  )
+  for _, paths in pairs(grouped_paths) do
+    local cloned_route = tablex.deepcopy(rs)
```

Comment:
I think the biggest concern I have about the PR is right here, since we know there are users out there with service/route counts in the mid 5 figures. I think it's a case where it's worth splitting some hairs to see what we can do to cut down on copying tables.

For one, I wonder if it's possible to get away with just copying the route and re-using the service?

```lua
local cloned_route = {
  route = tablex.deepcopy(rs.route),
  original_route = rs.route,
  service = rs.service,
}
cloned_route.route.paths = paths
cloned_route.route.id = utils.uuid()
```

Additionally, as long as we're reasonably certain that `atc` doesn't modify the input, I think you can likely use a [shallow clone](https://github.com/Kong/kong/blob/c0a7eb816dcdf02b5ea63767176d7420df1059ff/kong/tools/utils.lua#L612-L624) of the route rather than a deep copy. The route entity has a bunch of table fields, so a shallow clone will be a lot cheaper than a deep copy:

```lua
local cloned_route = {
  route = utils.shallow_copy(rs.route),
  original_route = rs.route,
  service = rs.service,
}
cloned_route.route.paths = paths
cloned_route.route.id = utils.uuid()
```

There is _some_ maintenance risk involved with this (what if a new property is added to this `rs` table?), but thorough test coverage should/would catch that.

Reply:
Thank you for your review and the optimization suggestion.

I've opted to check the `rs` table for unexpected content while processing it.  According to my reading, `atc` does not update the routes itself, so sharing should be fine and I've gone with the shallow copy.